### PR TITLE
Enable multi-party submissions [KVL-708]

### PR DESCRIPTION
**Pull request by rautenrieth-da**
This is the final PR that enables multi-party submissions in the ledger API.

## Summary

This PR enables multi-party submissions in the ledger API server. Once it's merged, multi-party submissions should be fully usable (supported on sandbox classic and all kvutils ledgers).

## Changelog

```
CHANGELOG_BEGIN
- [Ledger API] The ledger API now supports multi-party submissions.
  In order to use multi-party submissions, use the new act_as and
  read_as fields in submission requests.
CHANGELOG_END
```

## Checklist

The following needs to be done before this PR can be merged:

- [x] Rebase this branch onto master after KVL-707 is merged
- [x] Remove temporary helper functions (they should not be used anywhere after rebasing onto KVL-707):
  - `SubmitterInfo.singleSubmitterOrThrow`
  - `SubmitterInfo.withSingleSubmitter`
  - `TxEntry.singleSubmitterOrThrow`
- [x] Run the new `MultiPartySubmissionIT` for kvutils ledgers. To do that, search for `--include=ParticipantPruningIT` in `ledger/ledger-on-memory/BUILD.bazel` and `ledger/ledger-on-sql/BUILD.bazel`, duplicate the enclosing `conformance_test` block, and replace `ParticipantPruningIT` with `MultiPartySubmissionIT` in it.

### CODE REVIEW

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:209]: > also remove the call to `actAs.head` in validateCommands()

This was done in an earlier PR

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/MultiPartySubmissionIT.scala
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/MultiPartySubmissionIT.scala:174]: Juding from [here](https://github.com/digital-asset/daml/blob/master/language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/LFUtil.scala#L248) and [here](https://github.com/digital-asset/daml/blob/master/language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/TemplateCompanion.scala#L83), the actor argument in codegen exercise methods is not used for anything.

@S11001001 is that correct? If it's not used, do we keep that argument for backwards compatibility?

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (bame-da) [inline ledger/test-common/src/main/daml/model/Test.daml:617]: Together with the actual tests above, this covers the common uses of multi-party submissions, but only at the command level. `actAs` only has an effect at the command level as it sets the available authorizers for top-level actions, but `read` as has an effect on any action in the transaction tree.

For completeness, I would therefore add a few tests that test reads at least one level down in the transaction tree. For example:

1. Give the template a contract key `key this: MultiPartyContract; maintainer key.parties`.
2. Give the template three choices `FetchOther, FetchOtherByKey, LookupOther`, which take a cid, or key or a second contract instance, respectively.
3. Test all 12 combinations of
  - 3 choices
  - Well-authorized or not
  - Contract to be looked up visible or not

Comment (rautenrieth-da) [reply ledger/test-common/src/main/daml/model/Test.daml:617]: Is the following an example of how this should work? 

1. Alice creates a `MultiPartyContract` for herself (let's call it `cA`)
2. Bob creates a `MultiPartyContract` for himself (let's call it `cB`)
3. Bob exercise a `FetchOther` choice on his contract `cB`, which attempts to fetch contract `cA`. This is well-authorized, but the contract lookup will fail unless Alice is a `readAs` party or `cA` has been divulged to Bob.

Comment (cocreature) [reply ledger/test-common/src/main/daml/model/Test.daml:617]: https://github.com/digital-asset/daml/blob/70af09db51e9f0d2b5733f108e06e096526cf02f/compiler/damlc/tests/src/DA/Test/ScriptService.hs#L653 might be a reasonable example from the daml script tests.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala:40]: Will this become non-optional in the future?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:557]: This makes sense. Are we planning on removing the other `submitRequest` that just takes a single party at some point in the future?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:587]: Same question as above.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:565]: Consider using `.widen` instead of `.map(_.unwrap)` to avoid creating a new list.

**Review by SamirTalwar-DA — APPROVED** [review]

**Comment by SamirTalwar-DA** [comment]
I have approved the code changes, but please get approval from @bame-da before merging.

**Comment by kamil-da** [comment]
Thank you @SamirTalwar !

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala:40]: As far as I know it will become non-optional in the future. We keep this test as optional not to force all customers to implement multi-party submissions. Is that correct @gerolf-da ?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:557]: I think that these two variants do slightly different things because the single-party one uses `party` field, while the multi-party one uses `actAs` field. The `party` field is still present to provide backward compatibility, so it might be beneficial to have some tests covering that too. What do you think?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:565]: Can I use `.widen` when the type of element is changing? In this example `.unwrap` maps from `Party` to `String`. `.widen` would change the collection type but not the element type. Am I correct?

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (fabiotudone-da) [reply ledger/test-common/src/main/daml/model/Test.daml:617]: @kamil-da This seems to be resolved, isn't it?

**Review by fabiotudone-da — APPROVED** [review]
LGTM, same conditional as @SamirTalwar-DA's (see my question).

File: ledger/test-common/src/main/daml/model/Test.daml
Comment (kamil-da) [reply ledger/test-common/src/main/daml/model/Test.daml:617]: Yes, @rautenrieth-da created additional tests. (See: https://github.com/digital-asset/daml/pull/8266/commits/acd3745c6fc14d66fe3200d2b3b5fd660facd3d6)

**Comment by kamil-da** [comment]
Thank you @fabiotudone-da !

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala
Comment (S11001001) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:565]: Use `Party.unsubst` on the list; `widen` will not work. `map` is definitely unnecessary.

However in general @kamil-da `widen` _does_ change the element type. It's just the wrong tool for this particular situation.

Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:565]: I haven't noticed the subtype relation between `Party` and `String`. Thank you @S11001001 !

Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala]: Fixed.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala
Comment (gerolf-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala:40]: Yes, that's the idea behind it. We can make it non-optional in the final 1.9 release, but for now I'd like to not force non-KV ledgers to HAVE to implement it when updating to the next 1.9 SDK snapshot.

**Comment by gerolf-da** [comment]
@kamil-da: Please add a conformance test for DAML on SQL and sandbox classic as well.

**Comment by kamil-da** [comment]
@gerolf-da Do you mean adding `MultiPartySubmissionIT` as a conformance test to those packages? For the `daml-on-sql` it's been already added [here](https://github.com/digital-asset/daml/pull/8266/files#diff-a25906b22e1d3a3959e7089c2e83847e147df929f78f0efd701272d472c93fb6) - is that what you meant?

**Comment by gerolf-da** [comment]
> @gerolf-da Do you mean adding `MultiPartySubmissionIT` as a conformance test to those packages? For the `daml-on-sql` it's been already added [here](https://github.com/digital-asset/daml/pull/8266/files#diff-a25906b22e1d3a3959e7089c2e83847e147df929f78f0efd701272d472c93fb6) - is that what you meant?

Yikes, sorry. Of course it's already there. I was scrolling through too quickly and only parsed `ledger-on-sql` and `ledger-on-memory.

**Review by gerolf-da — APPROVED** [review]

**Comment by kamil-da** [comment]
Got a green light from @moritzkiefer-da and @bame-da .
🚀

**Pull request merged**
